ipatests/util.py

```diff
@@ -761,20 +761,25 @@ def change_principal(principal, password=None, client=None, path=None,
     client.Backend.rpcclient.disconnect()
 
     try:
-        with private_ccache(ccache_name):
-            if keytab:
-                kinit_keytab(principal, keytab, ccache_name)
-            else:
-                kinit_password(principal, password, ccache_name,
-                               canonicalize=canonicalize,
-                               enterprise=enterprise)
-            client.Backend.rpcclient.connect()
+        if keytab:
+            kinit_keytab(principal, keytab, ccache_name)
+        else:
+            kinit_password(principal, password, ccache_name,
+                           canonicalize=canonicalize,
+                           enterprise=enterprise)
+        client.Backend.rpcclient.connect(ccache=ccache_name)
 
-            try:
-                yield
-            finally:
-                client.Backend.rpcclient.disconnect()
+        try:
+            yield
+        finally:
+            client.Backend.rpcclient.disconnect()
     finally:
+        # If we generated a ccache name, try to remove it, but don't fail
+        if not path:
+            try:
+                os.remove(ccache_name)
+            except OSError:
+                pass
```

Comment:
IMO a better fix would be to let `private_ccache()` create and delete the temporary ccache rather than do it manually:
```python
    try:
        with private_ccache(path) as ccache_name:
            if keytab:
                kinit_keytab(principal, keytab, ccache_name)
            else:
                kinit_password(principal, password, ccache_name,
                               canonicalize=canonicalize,
                               enterprise=enterprise)
            client.Backend.rpcclient.connect(ccache=ccache_name)

            try:
                yield
            finally:
                client.Backend.rpcclient.disconnect()
    finally:
        client.Backend.rpcclient.connect()